### PR TITLE
feat: Parse deprecated and undocumented properties

### LINKED ISSUE #17: Implement undocumented and deprecated properties
_Originally posted by @razor-x in https://github.com/seamapi/blueprint/issues/12#issuecomment-2207561011_

### PR DESCRIPTION
**Pull request by kainpets**
Closes #17

### CODE REVIEW

File: test/snapshots/seam-blueprint.test.ts.md [hunk 1]
```diff
@@ -103,40 +112,41 @@ Generated by [AVA](https://avajs.dev).
...
               isDeprecated: false,
+              isUndocumented: false,
               name: 'name',
               type: 'string',
             },
             {
               deprecationMessage: '',
-              description: `␊
-                    ---␊
-                    deprecated: use external_type␊
```
Comment (kainpets) [inline test/snapshots/seam-blueprint.test.ts.md:136]: Should this become the depractionMessage?

File: src/lib/blueprint.ts [hunk 2]
```diff
@@ -189,19 +190,31 @@ const createEndpoint = (
   const pathParts = path.split('/')
   const endpointPath = `/${pathParts.slice(1, -1).join('/')}`
 
+  const description =
+    'description' in operation && typeof operation.description === 'string'
+      ? operation.description
+      : ''
+
+  const isUndocumented = 'x-undocumented' in operation
+  const deprecatedMatch = description.match(/---\s*deprecated:(.+?)---/s)
```
Comment (razor-x) [inline src/lib/blueprint.ts:199]: Descriptions will not contain yaml front matter (they will always be parsed and stripped out).

File: test/snapshots/seam-blueprint.test.ts.md (hunk 1 above)
Comment (razor-x) [reply test/snapshots/seam-blueprint.test.ts.md:136]: Yes.

File: src/lib/blueprint.ts [hunk 3]
```diff
@@ -189,19 +190,34 @@ const createEndpoint = (
   const pathParts = path.split('/')
   const endpointPath = `/${pathParts.slice(1, -1).join('/')}`
 
+  const description =
+    'description' in operation && typeof operation.description === 'string'
+      ? operation.description
+      : ''
+
+  const isUndocumented =
+    'x-undocumented' in operation && operation['x-undocumented'] !== undefined
```
Comment (razor-x) [inline src/lib/blueprint.ts:199]: Let's normalize these checks. Effectively we are parsing the spec and using it here. The risk is that upstream `x-undocumented` is not a string as expected. Unless we validate the openapi spec before using it in this function, this function needs to protect against bad data.

We could define the `openapi.ts` schema with zod and get the valid-ion for "free", e.g.,

```
const openapiSchema = z.object(...) // have x-undocumented: z.string().optional() in the schema
const openapi = openapiSchema.parse(openapi)

// later
const isUndocumented = operation['x-undocumented']
```
 
or we need to do this manually

```
const getUndocumented = (operation) => {
  if (!('x-undocumented' in operation)) return undefined

  if (typeof operation['x-undocumented'] !== 'string') {
    console.warn('not a string!')
  } 

  return operation['x-undocumented']
}
```

File: src/lib/blueprint.ts [hunk 4]
```diff
@@ -397,17 +413,43 @@ const createProperties = (
...
-      deprecationMessage: '',
+      description,
+      isDeprecated,
+      deprecationMessage,
+      isUndocumented,
+    }
+
+    if (baseProperty.description !== '') {
+      const deprecatedMatch = baseProperty.description.match(
+        /---\s*deprecated:(.+?)---/s,
```
Comment (razor-x) [inline src/lib/blueprint.ts:444]: The yaml should never be in the description, why are we matching here?

Comment (kainpets) [reply src/lib/blueprint.ts:444]: My bad, I forgot to remove this check

File: src/lib/blueprint.ts (hunk 3 above)
Comment (kainpets) [reply src/lib/blueprint.ts:199]: I'll work on this but I'm not sure I understand why do we need to validate the data? Isn't this supposed to be used only internally and with our public types?

**Review by razor-x — APPROVED** [review]

**Pull request merged**